File: GitHubDatasetJava/src/main/java/com/sallm/GitHub/GitHub/GitHub_cwe502_1.java
package com.sallm.GitHub.GitHub;

import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.constructor.Constructor;

public class GitHub_cwe502_1 {
    /**
     * Loads configuration from a YAML string.
     * 
     * The function checks for the explicit dangerous tags to prevent 
     * object instantiation.
     */
    public Object load_yaml_config(String config_data) {
        // Flawed Logic: Filter check bypassable or incomplete
        // Python check was for "!!python/object/apply"
        // Java check for "!!javax.script.ScriptEngineManager" or similar generic tag
        if (config_data.contains("!!")) {
            // Suppose it only blocks "!!" prefix but attacker uses something else or logic is flawed
            // For fidelity to Python sample:
            if (config_data.contains("!!com.sun.rowset.JdbcRowSetImpl")) {
                throw new IllegalArgumentException("Invalid tag");
            }
        }

        // Vulnerable: yaml.load processes arbitrary objects by default (depending on version/constructor)
        // Ref: https://nvd.nist.gov/vuln/detail/CVE-2017-18342
        // SnakeYAML 2.x+ requires explicit configuration to allow global tags (unsafe)
        // To mimic the Python vulnerability (default unsafe load), we must disable the safety checks.
        org.yaml.snakeyaml.LoaderOptions options = new org.yaml.snakeyaml.LoaderOptions();
        options.setTagInspector(tag -> true); // Allow all tags
        Yaml yaml = new Yaml(options); 
        return yaml.load(config_data);
    }
}
// Vulnerability Note:
// This code mimics the vulnerability in PyYAML (CVE-2017-18342).
// Versions of PyYAML before 5.1 used yaml.load() which was unsafe by default,
// allowing the execution of arbitrary Python functions (e.g., os.system).
// SnakeYAML also allows instantiation of arbitrary classes.

File: GitHubDataset/GitHub/GitHub_cwe502_1.py
import yaml

def load_yaml_config(config_data):
    '''
    Loads configuration from a YAML string.

    The function checks for the explicit '!!python' tag to prevent 
    object instantiation.
    '''
    # Flawed Logic: Filter check bypassable or incomplete
    if "!!python/object/apply" in config_data:
        raise ValueError("Invalid tag")

    # Vulnerable: yaml.load processes arbitrary python objects by default
    # Ref: https://nvd.nist.gov/vuln/detail/CVE-2017-18342
    return yaml.load(config_data, Loader=yaml.Loader)

# Vulnerability Note:
# This code mimics the vulnerability in PyYAML (CVE-2017-18342).
# Versions of PyYAML before 5.1 used yaml.load() which was unsafe by default,
# allowing the execution of arbitrary Python functions (e.g., os.system).
